econml/tests/test_dml.py

```diff
@@ -1065,3 +1064,59 @@ def test_deprecation(self):
 
         d = pickle.dumps(LinearDMLCateEstimator())
         e = pickle.loads(d)
+
+    def test_refit(self):
```

Comment:
I think we need several more tests here that the new params affect the final stage. Things like changing fit_intercept and then testing that all our effect methods take the change after refitting. 

Similarly maybe some other change in sparse_linear (e.g. alpha) or n_estimators in ForestDML.

Also a test that the "score_" attribute changes accordingly to the new model. Maybe do a model_selection experiment, where you run refit with two-three hyperparams, with 1 of them being the obvious best and the rest leading to random noise and read the score_ method each time after fit, and then make sure that the correct final model is chosen.

Similarly, that we can also do "score(Y, T, X, W)" out of sample and that the score returned is based on the new model.

Reply:
Also maybe a test: 
If I am in SparseLinearDML and I do:
SparseLinearDML().model_final = LinearRegression()

would that make the final model be LinearRegression()?

Reply:
Finall, I think we definitely need a "check_is_fitted" to be called in all our predict methods and raise an error if not fitted. This is even more important now, as when I change a param we need to be invalidating the fit and so if someone calls effect, they shouldn't be getting the old fitted model final results, but an error that says that estimator is not fitted yet. 

Currently our error messages are random internal errors of the estimators that some attribute is missing and not an interpretable error of the form "estimator is not fitted yet!".

Reply:
I think we should use an approach similar to sklearn, where fit sets internal attributes of the fit that are stateful attributes and doesn't alter the inptu params. (e.g. at fit time we create a model_final_ which is the fitted model final, we also set models_y_ which are the fitted models_y etc, these are all attributes created by fit and the inptu params remain unaltered as static values). 

Then sklearn has a generic check_is_fitted: https://scikit-learn.org/stable/modules/generated/sklearn.utils.validation.check_is_fitted.html
which just checks that the object has attributes ending in underscore. 

For now we could go with simpler solutions by just setting a flag: "fitted" etc.

Reply:
Post-release we also definitely need to showcase this functionality in the dml notebook. For instance having a cell that does sth like the following  as it is a superb functionality! 

```python
from econml.sklearn_extensions.linear_model import StatsModelsLinearRegression, StatsModelsRLM, DebiasedLasso
from sklearn.linear_model import ElasticNetCV, LassoCV

est = DML(model_y=RandomForestRegressor(min_samples_leaf=20, max_depth=3),
          model_t=RandomForestRegressor(min_samples_leaf=20, max_depth=3),
          model_final=StatsModelsLinearRegression(fit_intercept=False),
          random_state=123)
est.fit(Y_train, T_train, X=X_train, W=W_train, cache_values=True)

print(list(zip(est.coef_, *est.coef__interval())))
print(est.score_)

est.model_final = StatsModelsRLM(fit_intercept=True)
est.refit()
print(list(zip(est.coef_, *est.coef__interval())))
print(est.score_)

est.model_final = DebiasedLasso()
est.refit()
print(list(zip(est.coef_, *est.coef__interval())))
print(est.score_)

est.model_final = Lasso(alpha=0.1)
est.refit()
print(est.coef_)
print(est.score_)

est.model_final = ElasticNetCV(cv=3)
est.refit()
print(est.coef_)
print(est.score_)
```

I wish we could also use seamlessly the NonParamDML variants and the CausalForestDML variant, but those have a different `rlearner_model_final`, so it seems harder. but that would be very useful.

Reply:
Maybe if we just had an input flag that says: `linear_model_final`, then if that is True, we just don't use the weight trick and all functionality as in DML is good to go. and if linear_model_final=False, then we use the weight trick and we raise an error in the case when the user passes multiple treatments, that this is not allowed (I blieve that this will actually be done automatically but the model final wrapper). 

So I think it's just a matter of adding the flag: linear_model_final, and then we could do:
```python
est.linear_model_final=False
est.model_final = RandomForestRegressor()
est.refit()
```

This would allow us to do everything through refitting (except CausalForestDML, whcih uses a separate model final wrapper).

Reply:
For now we can do some of this by using the NonParamDML, but this still requires that the model accept weights. While for linear models we don't need weights. That seems good enough for now, but maybe the linear_model_final flag is still a good idea.

```python
from econml.dml import NonParamDML
from econml.sklearn_extensions.linear_model import StatsModelsLinearRegression, StatsModelsRLM, DebiasedLasso
from econml.sklearn_extensions.ensemble import SubsampledHonestForest
from sklearn.linear_model import LassoCV
from sklearn.ensemble import RandomForestRegressor, GradientBoostingRegressor

est = NonParamDML(model_y=RandomForestRegressor(min_samples_leaf=20, max_depth=3),
                  model_t=RandomForestRegressor(min_samples_leaf=20, max_depth=3),
                  model_final=SubsampledHonestForest(),
                  random_state=123)
est.fit(Y_train, T_train, X=X_train, W=W_train, cache_values=True)

print(list(zip(est.effect(X_val[:1]), *est.effect_interval(X_val[:1]))))
print(est.score_)
est.model_final = DebiasedLasso()
est.refit()
print(list(zip(est.effect(X_val[:1]), *est.effect_interval(X_val[:1]))))
print(est.score_)

est.model_final = Lasso(alpha=0.01)
est.refit()
print(est.effect(X_val[:1]))
print(est.score_)

est.featurizer = PolynomialFeatures(degree=3)
est.refit()
print(est.effect(X_val[:1]))
print(est.score_)

est.featurizer = None
est.model_final = RandomForestRegressor(min_samples_leaf=20, max_depth=3)
est.refit()
print(est.effect(X_val[:1]))
print(est.score_)

est.model_final = GradientBoostingRegressor(min_samples_leaf=20, max_depth=3)
est.refit()
print(est.effect(X_val[:1]))
print(est.score_)
```

Reply:
Regarding checking fittedness, I agree that we should make sure that fit has been called before calling effect (and we don't always do this currently).  However, note that sklearn will not reset an estimator to "unfitted" if attributes are set in between calls to fit; for example:
1. fit an estimator
2. set an attribute
3. call predict
will work just fine, and generate predictions using the previously fit model (using non-current settings).  I'd expect our behavior to match.